File: py/prob_48.py
class Solution:
    def rotate(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: void Do not return anything, modify matrix in-place instead.
        """
        N = len(matrix)
        
        # Reduce the matrix diagonally
        for i in range(int(N/2)):
            
            # Go thorugh every node at edge
            for j in range(i, N-i-1):
                copy_curr = matrix[i][j]
                copy_next = None 
                
                # Perform rotation cycle
                for k in range(4):
                    t = i
                    i = j
                    j = N - t - 1
                    copy_next = matrix[i][j]
                    matrix[i][j] = copy_curr
                    copy_curr = copy_next
        
        return

File: cpp/prob_48.cpp
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        int N = matrix.size();
        
        for (int i=0; i < (int) N/2; i++) {
            
            for (int j=i; j < N-i-1; j++) {
                int copy_curr = matrix[i][j];
                int copy_next = 0;
                
                for (int k=0; k < 4; k++) {
                    int t = i;
                    i = j;
                    j = N-t-1;
                    copy_next = matrix[i][j];
                    matrix[i][j] = copy_curr;
                    copy_curr = copy_next;
                }
            }
        }
        
        return;
    }
};